File: api/v1/customer.py
"""客户管理 API。

路由结构：
- 读端点 (`GET /customers`)：MANAGER + CLERK + CNC_PROGRAMMER 共享，
  文员/编程员要拉客户列表做筛选。
- CRUD 端点（POST / 单条 GET / update / soft-delete）：MANAGER + CLERK。
  与申请人目录一致（user 拍板"MANAGER + CLERK 都可写"）。

两个 router 共用 `/customers` 前缀；由于 FastAPI 的依赖是按路由粒度生效，
两者并列注册即可。
"""
from fastapi import APIRouter, Depends, status as http_status

from api.deps import get_customer_service
from core.permission import require_role, require_roles
from model.enums import UserRole
from schema.customer import (
    CustomerCreateRequest,
    CustomerOut,
    CustomerUpdateRequest,
)
from service import CustomerService

# ============================================================
# 读路由：开放给 MANAGER + CLERK + CNC_PROGRAMMER
# ============================================================
read_router = APIRouter(
    prefix="/customers",
    tags=["客户管理"],
    dependencies=[
        Depends(require_roles(
            UserRole.MANAGER, UserRole.CLERK, UserRole.CNC_PROGRAMMER,
        ))
    ],
)


@read_router.get(
    "",
    response_model=list[CustomerOut],
    summary="客户列表（含 parent_name，前端按 parent_name/child_name 唯一定位）",
)
async def list_customers(
    svc: CustomerService = Depends(get_customer_service),
) -> list[CustomerOut]:
    return await svc.list_customers()


@read_router.get(
    "/{customer_id}",
    response_model=CustomerOut,
    summary="客户详情",
)
async def get_customer(
    customer_id: str,
    svc: CustomerService = Depends(get_customer_service),
) -> CustomerOut:
    return await svc.get_customer(customer_id)


# ============================================================
# 写路由：仅 MANAGER + CLERK
# ============================================================
write_router = APIRouter(
    prefix="/customers",
    tags=["客户管理(写)"],
    dependencies=[
        Depends(require_roles(UserRole.MANAGER, UserRole.CLERK))
    ],
)


@write_router.post(
    "",
    response_model=CustomerOut,
    status_code=http_status.HTTP_201_CREATED,
    summary="新增客户（parent_id 留空=一级客户）",
)
async def create_customer(
    payload: CustomerCreateRequest,
    svc: CustomerService = Depends(get_customer_service),
) -> CustomerOut:
    return await svc.create_customer(payload)


@write_router.post(
    "/{customer_id}/update",
    response_model=CustomerOut,
    summary="更新客户字段",
)
async def update_customer(
    customer_id: str,
    payload: CustomerUpdateRequest,
    svc: CustomerService = Depends(get_customer_service),
) -> CustomerOut:
    return await svc.update_customer(customer_id, payload)


@write_router.post(
    "/{customer_id}/soft-delete",
    summary="软删客户（有子节点或被引用时拒绝）",
)
async def soft_delete_customer(
    customer_id: str,
    svc: CustomerService = Depends(get_customer_service),
) -> dict:
    await svc.soft_delete_customer(customer_id)
    return {"ok": True}

File: frontend/src/api/customer.ts
// 客户 API（走 @/api/http 统一 axios 客户端）。
// Excel 批量导入零件时用 listCustomers() 拉全量客户，
// 客户端按 `parent_name / name` 唯一定位叶子客户的 id。

import { api } from '@/api/http'

export interface Customer {
  id: string
  name: string
  parent_id: string | null
  /** 仅叶子节点的二级客户有 parent_name；一级节点为 null。 */
  parent_name: string | null
  /**
   * 一级客户的序列号前缀（A-Z 单字符）；叶子客户继承父，本字段恒为 null。
   * 后端 schema 端为 `str | None`，前端保持 `string | null` 与 schema 一致。
   */
  serial_prefix: string | null
}

export interface CustomerCreatePayload {
  name: string
  /** null = 一级客户；非空 = 挂在某个一级客户下的二级客户。 */
  parent_id: string | null
  /**
   * 一级客户必填（A-Z 单字符）；叶子客户忽略本字段（永远继承父）。
   * 传 null = 不传（后端会拒）。
   */
  serial_prefix: string | null
}

export interface CustomerUpdatePayload {
  name?: string
  parent_id?: string | null
  /**
   * 一级客户的序列号前缀（A-Z 单字符）。后端 schema 无 "传 null = 清空"
   * 语义，所以本字段 None 即"不改"；想改时直接传新字母。
   * 叶子客户忽略本字段。
   */
  serial_prefix?: string | null
}

export async function listCustomers(): Promise<Customer[]> {
  const resp = await api.get<Customer[]>('/customers')
  return resp.data
}

export async function getCustomer(id: string): Promise<Customer> {
  const resp = await api.get<Customer>(`/customers/${id}`)
  return resp.data
}

export async function createCustomer(
  payload: CustomerCreatePayload,
): Promise<Customer> {
  const resp = await api.post<Customer>('/customers', payload)
  return resp.data
}

export async function updateCustomer(
  id: string,
  payload: CustomerUpdatePayload,
): Promise<Customer> {
  const resp = await api.post<Customer>(`/customers/${id}/update`, payload)
  return resp.data
}

export async function softDeleteCustomer(id: string): Promise<void> {
  await api.post(`/customers/${id}/soft-delete`)
}